### PR TITLE
Parse acpi tables

### PR DESCRIPTION
**Pull request by Jett59**
This adds code to parse all of the important (as far as I can predict) ACPI tables for kernel initialization. On aarch64, this means the XSDT/RSDT, MADT, FADT and GTDT. On x86_64, it is the XSDT/RSDT, MADT and HPET.

So far, all it does with the information is print it to the screen. The next thing to do will be to make it set up timers and (eventually) initialize secondary processors.

### CODE REVIEW

File: kernel/src/acpi/madt.rs [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+macro_rules! default_operator {
+    ($left:expr, $default:tt|$given:tt, $right:expr) => {
+        $left $given $right
+    };
+    ($left:expr, $default:tt, $right:expr) => {
+        $left $default $right
+    };
+}
+
+macro_rules! madt_entry {
```
Comment (luketn) [inline kernel/src/acpi/madt.rs:62]: I'm not sure that these macro rules are making the code more readable. Is it possible to do this size comparison without the custom macro?

File: kernel/src/heap.rs
Comment (luketn) [inline kernel/src/heap.rs:1]: Nice - Seems like we are improving memory management and handling some cases we weren't before. I like the code comments too. Especially for complex things like memory allocations these comments may be important in the future. 
I'd suggest adding some unit tests here too - perhaps for things like checking whether the memory is being allocated on a boundary and making the appropriate adjustments. You could probably isolate those checks to a function and test that (independent of any actual memory).

File: kernel/src/x86_64/acpi/hpet.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    access_width: u8,
+    address: u64,
+    hpet_number: u8,
+    minimum_tick: u16,
+    page_protection: u8,
+}
+
+impl Validateable for HpetTableBody {
+    fn validate(&self) -> bool {
+        true // TODO: Actually validate this
```
Comment (luketn) [inline kernel/src/x86_64/acpi/hpet.rs:23]: Maybe worth doing that before moving on?

**Review by luketn — CHANGES_REQUESTED** [review]
Code changes look good. A couple of suggestions. But when I tried to run it locally I got this error on QEMU:

```
Initialized
the display (obviously)
Acpi tables at address 0x43a3fe98
Kernel panic：
panicked at src/acpi.rs:165:42：
misaligned
pointer dereference: address
multiple of a×e but
⑦×£+££a0000042¢ebc
```

**Comment by Jett59** [comment]
I couldn't reproduce the panick on my machine. Maybe it's something to do with qemu version? Also, I think some of the text was cut off in the error message.

File: kernel/src/acpi.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+#[derive(Debug)]
+pub enum AcpiTableParseError {
+    InvalidHeader,
+    ChecksumFailure,
+}
+
+impl AcpiTableHandle {
+    /// # Safety
+    /// If the provided address doesn't refer to an ACPI table, there will likely be undefined behaviour since it is impossible to tell at this point.
+    /// Additionally, it is UB if the table has already been mapped, since this would create possible aliasing issues.
```
[the remark this answers is not in the document]
Comment (Jett59) [reply kernel/src/acpi.rs:68]: A common abbreviation for 'Undefined Behaviour'.

File: kernel/src/acpi/madt.rs (hunk 1 above)
Comment (Jett59) [reply kernel/src/acpi/madt.rs:62]: Yes, but it would be exceedingly repetative. Every single one would have to have the structure itself, plus an impl block for the getter methods, plus another impl block for Validatable. But is it just the => true but that you don't like? Because we could easily just remove it. I added that part without really considering that there would probably be no places where we could reasonably validate anything. Without that bit it would just be a simple Rust structure definition wrapped in the macro, possibly with the comparison operator attribute. I suppose we could remove that as well and just not use the macro when we need to have a different operator as well.
